src/storage/temp_table.cpp

```diff
@@ -0,0 +1,133 @@
+
+//===----------------------------------------------------------------------===//
+//
+//                         PelotonDB
+//
+// temp_table.cpp
+//
+// Identification: /peloton/src/storage/temp_table.cpp
+//
+// Copyright (c) 2015, 2016 Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "storage/temp_table.h"
+
+#include "catalog/schema.h"
+#include "common/exception.h"
+#include "common/logger.h"
+#include "storage/tile_group.h"
+#include "storage/tile_group_header.h"
+#include "storage/tuple.h"
+
+namespace peloton {
+namespace storage {
+
+TempTable::TempTable(const oid_t &table_oid, catalog::Schema *schema,
+                     const bool own_schema)
+    : AbstractTable(table_oid, schema, own_schema) {
+  // We only want to instantiate a single TileGroup
+  AddDefaultTileGroup();
+}
+
+TempTable::~TempTable() {
+  // Nothing to see, nothing to do
+}
+
+ItemPointer TempTable::InsertTuple(const Tuple *tuple,
+                                   concurrency::Transaction *transaction,
+                                   ItemPointer **index_entry_ptr) {
+  PL_ASSERT(tuple != nullptr);
+  PL_ASSERT(transaction == nullptr);
+  PL_ASSERT(index_entry_ptr == nullptr);
+
+  std::shared_ptr<storage::TileGroup> tile_group;
+  oid_t tuple_slot = INVALID_OID;
+  oid_t tile_group_id = INVALID_OID;
+
+  for (int i = 0, cnt = (int)tile_groups_.size(); i < cnt; i++) {
+    tile_group = tile_groups_[i];
+    tuple_slot = tile_group->InsertTuple(tuple);
+
+    // now we have already obtained a new tuple slot.
+    if (tuple_slot != INVALID_OID) {
+      tile_group_id = tile_group->GetTileGroupId();
+      LOG_TRACE("Inserted tuple into %s", GetName().c_str());
+      break;
+    }
+  }
+  // if this is the last tuple slot we can get
+  // then create a new tile group
+  if (tuple_slot == tile_group->GetAllocatedTupleCount() - 1) {
+    AddDefaultTileGroup();
+  }
+  if (tile_group_id == INVALID_OID) {
+    LOG_WARN("Failed to get tuple slot.");
+    return INVALID_ITEMPOINTER;
+  }
+
+  // Set tuple location and increase our counter
+  ItemPointer location(tile_group_id, tuple_slot);
+  IncreaseTupleCount(1);
+
+  // Make sure that we mark the tuple as active in the TileGroupHeader too
+  // If you don't do this, then anybody that tries to use a LogicalTile wrapper
+  // on this TempTable will not have any active tuples.
+  auto tile_group_header = tile_group->GetHeader();
+  tile_group_header->SetTransactionId(location.offset, INITIAL_TXN_ID);
```

Comment:
I think it's not enough to just set the transaction id field in the header. For insert executor, it will later call the `TransactionManager.PerformInsert()` to set up the header of a newly inserted tuple. Here you may need to set the begin timestamp of this tuple as MIN_CID and the end timestamp as MAX_CID to ensure it pass the visibility test in the upper layer.